File: swift/Sources/TesseraCore/OTP.swift
#if canImport(CryptoKit)
import CryptoKit
#else
import Crypto
#endif
import Foundation

/// HOTP (RFC 4226), TOTP (RFC 6238), and the Steam Guard variant, built on
/// swift-crypto / CryptoKit HMAC. Pinned against the shared interop vectors.
public enum OTP {
    public enum Algorithm: String, Sendable {
        case sha1 = "SHA1", sha256 = "SHA256", sha512 = "SHA512"

        public static func parse(_ s: String) throws -> Algorithm {
            switch s.uppercased() {
            case "", "SHA1": return .sha1
            case "SHA256": return .sha256
            case "SHA512": return .sha512
            default: throw AccountError.invalid("unsupported algorithm \(s)")
            }
        }
    }

    private static func hmac(_ secret: Data, _ message: Data, _ alg: Algorithm) -> Data {
        let key = SymmetricKey(data: secret)
        switch alg {
        case .sha1:   return Data(HMAC<Insecure.SHA1>.authenticationCode(for: message, using: key))
        case .sha256: return Data(HMAC<SHA256>.authenticationCode(for: message, using: key))
        case .sha512: return Data(HMAC<SHA512>.authenticationCode(for: message, using: key))
        }
    }

    private static func truncate(_ mac: Data) -> UInt32 {
        let bytes = [UInt8](mac)
        let offset = Int(bytes[bytes.count - 1] & 0x0f)
        return (UInt32(bytes[offset] & 0x7f) << 24)
            | (UInt32(bytes[offset + 1]) << 16)
            | (UInt32(bytes[offset + 2]) << 8)
            | UInt32(bytes[offset + 3])
    }

    private static func counterBytes(_ counter: UInt64) -> Data {
        var c = counter.bigEndian
        return withUnsafeBytes(of: &c) { Data($0) }
    }

    public static func hotp(secret: Data, counter: UInt64, digits: Int, algorithm: Algorithm) throws -> String {
        if secret.isEmpty { throw AccountError.invalid("empty secret") }
        if digits < 6 || digits > 8 { throw AccountError.invalid("digits must be 6-8") }
        let mac = hmac(secret, counterBytes(counter), algorithm)
        let bin = truncate(mac)
        let mod = UInt32(pow(10.0, Double(digits)))
        return String(format: "%0\(digits)u", bin % mod)
    }

    public static func totp(secret: Data, time: Date, period: Int, digits: Int, algorithm: Algorithm) throws -> String {
        if period <= 0 { throw AccountError.invalid("period must be positive") }
        let counter = UInt64(Int64(time.timeIntervalSince1970) / Int64(period))
        return try hotp(secret: secret, counter: counter, digits: digits, algorithm: algorithm)
    }

    public static func remainingSeconds(time: Date, period: Int) -> Int {
        if period <= 0 { return 0 }
        return period - Int(Int64(time.timeIntervalSince1970) % Int64(period))
    }

    private static let steamAlphabet = Array("23456789BCDFGHJKMNPQRTVWXY")

    public static func decodeSteamSecret(_ b64: String) throws -> Data {
        guard let d = Data(base64Encoded: b64.trimmingCharacters(in: .whitespacesAndNewlines)) else {
            throw AccountError.invalid("steam secret base64")
        }
        return d
    }

    public static func steam(secret: Data, time: Date) throws -> String {
        if secret.isEmpty { throw AccountError.invalid("empty secret") }
        let counter = UInt64(Int64(time.timeIntervalSince1970) / 30)
        var full = truncate(hmac(secret, counterBytes(counter), .sha1))
        var out = ""
        for _ in 0..<5 {
            out.append(steamAlphabet[Int(full % UInt32(steamAlphabet.count))])
            full /= UInt32(steamAlphabet.count)
        }
        return out
    }

    /// Compute the current code for an account.
    public static func code(for a: Account, at time: Date) throws -> String {
        let alg = try Algorithm.parse(a.algorithm)
        switch a.type {
        case .totp: return try totp(secret: a.secret, time: time, period: a.period, digits: a.digits, algorithm: alg)
        case .steam: return try steam(secret: a.secret, time: time)
        case .hotp: return try hotp(secret: a.secret, counter: UInt64(a.counter), digits: a.digits, algorithm: alg)
        }
    }
}

File: go/internal/otp/otp.go
// Package otp implements RFC 4226 (HOTP), RFC 6238 (TOTP), and the Steam Guard
// variant from the Go standard library only. Keeping the security-critical code
// dependency-free lets it be pinned byte-for-byte against the Swift core via the
// shared interop vectors.
package otp

import (
	"crypto/hmac"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/base64"
	"encoding/binary"
	"fmt"
	"hash"
	"math"
	"strings"
	"time"
)

// Algorithm selects the HMAC hash used by HOTP/TOTP.
type Algorithm int

const (
	SHA1 Algorithm = iota
	SHA256
	SHA512
)

func (a Algorithm) String() string {
	switch a {
	case SHA1:
		return "SHA1"
	case SHA256:
		return "SHA256"
	case SHA512:
		return "SHA512"
	default:
		return "UNKNOWN"
	}
}

func (a Algorithm) new() func() hash.Hash {
	switch a {
	case SHA256:
		return sha256.New
	case SHA512:
		return sha512.New
	default:
		return sha1.New
	}
}

// ParseAlgorithm maps an otpauth algorithm string (case-insensitive) to an Algorithm.
func ParseAlgorithm(s string) (Algorithm, error) {
	switch strings.ToUpper(strings.TrimSpace(s)) {
	case "", "SHA1":
		return SHA1, nil
	case "SHA256":
		return SHA256, nil
	case "SHA512":
		return SHA512, nil
	default:
		return 0, fmt.Errorf("otp: unsupported algorithm %q", s)
	}
}

// dt performs RFC 4226 dynamic truncation, returning the 31-bit integer.
func dt(secret []byte, counter uint64, alg Algorithm) []byte {
	var buf [8]byte
	binary.BigEndian.PutUint64(buf[:], counter)
	mac := hmac.New(alg.new(), secret)
	mac.Write(buf[:])
	return mac.Sum(nil)
}

func truncate(sum []byte) uint32 {
	offset := sum[len(sum)-1] & 0x0f
	return (uint32(sum[offset]&0x7f) << 24) |
		(uint32(sum[offset+1]) << 16) |
		(uint32(sum[offset+2]) << 8) |
		uint32(sum[offset+3])
}

// HOTP computes the RFC 4226 code for the given counter. digits must be 6-8.
func HOTP(secret []byte, counter uint64, digits int, alg Algorithm) (string, error) {
	if len(secret) == 0 {
		return "", fmt.Errorf("otp: empty secret")
	}
	if digits < 6 || digits > 8 {
		return "", fmt.Errorf("otp: digits must be 6-8, got %d", digits)
	}
	bin := truncate(dt(secret, counter, alg))
	mod := uint32(math.Pow10(digits))
	return fmt.Sprintf("%0*d", digits, bin%mod), nil
}

// TOTP computes the RFC 6238 code at time t with T0=0.
func TOTP(secret []byte, t time.Time, period, digits int, alg Algorithm) (string, error) {
	if period <= 0 {
		return "", fmt.Errorf("otp: period must be positive, got %d", period)
	}
	counter := uint64(t.Unix() / int64(period))
	return HOTP(secret, counter, digits, alg)
}

// RemainingSeconds returns how many seconds the current TOTP step is still valid.
func RemainingSeconds(t time.Time, period int) int {
	if period <= 0 {
		return 0
	}
	return period - int(t.Unix()%int64(period))
}

const steamAlphabet = "23456789BCDFGHJKMNPQRTVWXY"

// DecodeSteamSecret decodes a Steam shared_secret (base64) to raw key bytes.
func DecodeSteamSecret(b64 string) ([]byte, error) {
	b, err := base64.StdEncoding.DecodeString(strings.TrimSpace(b64))
	if err != nil {
		return nil, fmt.Errorf("otp: steam secret base64: %w", err)
	}
	return b, nil
}

// Steam computes the 5-character Steam Guard code (HMAC-SHA1, period 30).
func Steam(secret []byte, t time.Time) (string, error) {
	if len(secret) == 0 {
		return "", fmt.Errorf("otp: empty secret")
	}
	counter := uint64(t.Unix() / 30)
	full := truncate(dt(secret, counter, SHA1))
	var b [5]byte
	for i := 0; i < 5; i++ {
		b[i] = steamAlphabet[full%uint32(len(steamAlphabet))]
		full /= uint32(len(steamAlphabet))
	}
	return string(b[:]), nil
}
